[3.x] Linux: Cache TTS voice list.

novalis opened the pull request:
Cherry-pick from 4.x.  Note that this simply removes the iteration through the voice list.  In 3.6 we do not use the language to do segmentation, so we do not need it at all (although later, it might be nice to pass it through to spd_set_language; it's unclear to me what precise effect this would have if we are already using a voice of the appropriate language).

Backport of #77767.

Thanks to @bruvzg for the really quick fix to #77754.

lawnjelly commented on platform/x11/tts_linux.cpp:
```diff
@@ -166,20 +171,15 @@ Array TTS_Linux::get_voices() const {
+		voice_d["name"] = E.key();
+		voice_d["id"] = E.key();
+		voice_d["language"] = E.get().language + "_" + E.get().variant;
+		list.push_back(voice_d);
```
If this hashmap is never used, can't you just cache the Array directly instead of the hashmap?

novalis replied:
If someone chooses to backport the language-specific word splitting, then they will want this.   This also matches up better with the 4.x code, making it easier to backport future bug fixes.  

If you want, I can change it, but I would generally prefer to keep the 3.x code similar to the 4.x code, rather than focusing on a small efficiency.

lawnjelly replied:
imo it depends how often this function `get_voices()` is called. If it is a one off, then there's no great problem - but in that case, there would be no need to cache you would have thought.

The main concern here is that it is building a new Array of Dictionaries each time `get_voices()` is called, where it could just cache the Array.

But looking at the linked issue, #77754 , the problem in 4.x seems to be that loading happens each time a speech event occurs, which is not the case in 3.x. So if `get_voices()` was only a one off (I have no idea if this is the case), there wouldn't seem much benefit to caching. :thinking:

novalis committed:
[Linux/TTS] Remove unused code

In 4.x, we need to find out what language is being used, so that
we can break words.  In 3.x, we assume that spaces spearate words.

So we do not need to find the language for the current voice.

akien-mga commented:
Thanks!
